src/v/kafka/security/scram_algorithm.h

```diff
@@ -0,0 +1,306 @@
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+#include "bytes/bytes.h"
+#include "hashing/secure.h"
+#include "kafka/logger.h"
+#include "random/generators.h"
+
+#include <absl/container/node_hash_map.h>
+
+#include <regex>
+
+/**
+ * scram algorthm - https://tools.ietf.org/html/rfc5802
+ *
+ * The implementation follows closely the rfc, with some scenarios and
+ * configuration not handled by kafka omitted from the implementation. The kafka
+ * implementation is closely based on the rfc as well, so they are fairly easy
+ * to compare.
+ *
+ * TODO
+ * ====
+ *
+ * 1. unnecessary copying from std::array<char> for hash digest types and
+ * built-in bytes type. the conversion is mostly mechanical, but std::array
+ * needs to be propogated to all other types which need to be templated on the
+ * same scram_algorithm template parameters.
+ */
+namespace kafka {
+
+class scram_exception final : public std::exception {
+public:
+    explicit scram_exception(ss::sstring msg) noexcept
+      : _msg(std::move(msg)) {}
+
+    const char* what() const noexcept final { return _msg.c_str(); }
+
+private:
+    ss::sstring _msg;
+};
+
+class scram_credential {
+public:
+    scram_credential() noexcept = default;
+
+    scram_credential(
+      bytes salt, bytes server_key, bytes stored_key, int iterations) noexcept
+      : _salt(std::move(salt))
+      , _server_key(std::move(server_key))
+      , _stored_key(std::move(stored_key))
+      , _iterations(iterations) {}
+
+    const bytes& salt() const { return _salt; }
+    const auto& server_key() const { return _server_key; }
+    const auto& stored_key() const { return _stored_key; }
+    int iterations() const { return _iterations; }
+
+private:
+    friend std::ostream& operator<<(std::ostream&, const scram_credential&);
+
+    bytes _salt;
+    bytes _server_key;
+    bytes _stored_key;
+    int _iterations{0};
+};
+
+/**
+ * First message received by the server.
+ */
+class client_first_message {
+public:
+    explicit client_first_message(bytes_view data);
+
+    client_first_message(client_first_message&&) = delete;
+    client_first_message& operator=(client_first_message&&) = delete;
+    client_first_message(const client_first_message&) = delete;
+    client_first_message& operator=(const client_first_message&) = delete;
+    ~client_first_message() noexcept = default;
+
+    const ss::sstring& authzid() const { return _authzid; }
+    const ss::sstring& username() const { return _username; }
+    ss::sstring username_normalized() const;
+    const ss::sstring& nonce() const { return _nonce; }
+
+    ss::sstring bare_message() const {
+        return fmt::format("n={},r={}", _username, _nonce);
+    }
+
+    bool token_authenticated() const;
+
+private:
+    friend std::ostream& operator<<(std::ostream&, const client_first_message&);
+
+    ss::sstring _authzid;
+    ss::sstring _username;
+    ss::sstring _nonce;
+    absl::node_hash_map<ss::sstring, ss::sstring> _extensions;
```

Comment:
It looks like only one extension value is used, so we can use one boolean flag (boo _is_token_authenticated) instead of the map.

Reply:
currently, yeh. i think there are additional flags that we may want to add.